util/circular_buffer.c

```diff
@@ -141,105 +147,183 @@ int32_t cb_remove(struct circular_buffer *desc)
 /**
  * @brief Get the number of elements in the buffer
  * @param desc - Circular buffer reference
- * @param nb_elements - Where to store the number of elements
+ * @param size - Where to store size of data available to read
  * @return
- *  - \ref SUCCESS : On success
- *  - \ref FAILURE : Otherwise
+ *  - \ref SUCCESS   - No errors
+ *  - -EINVAL   - Wrong parameters used
+ *  - -EOVERRUN - A buffer overrun occurred
  */
-int32_t cb_size(struct circular_buffer *desc, uint32_t *nb_elements)
+int32_t cb_size(struct circular_buffer *desc, uint32_t *size)
 {
-	if (!desc)
-		return FAILURE;
+	uint32_t nb_spins;
 
-	*nb_elements =  desc->nb_elem;
+	if (!desc || !size)
+		return -EINVAL;
+
+	if (desc->write.spin_count > desc->read.spin_count)
+		nb_spins = desc->write.spin_count - desc->read.spin_count;
+	else
+		/* Integer overflow on desc->write.spin_count */
+		nb_spins = UINT32_MAX - desc->read.spin_count +
+			   desc->write.spin_count + 1;
+
+	if (nb_spins > 0)
+		*size = desc->size + desc->write.idx - desc->read.idx;
+	else
+		*size = desc->write.idx - desc->read.idx;
+
+	if (*size > desc->size) {
+		*size = desc->size;
+		return -EOVERRUN;
+	}
 
 	return SUCCESS;
 }
 
-/**
- * @brief Write data to the buffer
- * @param desc - Circular buffer reference
- * @param data - Buffer from where data is copied to the circular buffer
- * @param nb_elements - Number of elements to be copied
- * @return
- *  - \ref SUCCESS : On success
- *  - \ref FAILURE : If overflow, if nb_elements == 0 or if desc is NULL
- */
-int32_t cb_write(struct circular_buffer *desc, const void *data,
-		 uint32_t nb_elements)
+static int32_t cb_prepare_async_operation(struct circular_buffer *desc,
```

Comment:
Maybe a short comment about why this function is needed will help with future reference.

Reply:
done